File: fibonacci/js/forwardRecursiveWithReduce.js
const fib = (n) => {
  if (n < 0 || !Number.isInteger(n)) {
    throw new Error('Invalid Input!')
  } else if (n < 2) {
    return n
  } else {
    const last2Fibs = new Array(n).fill(1).reduce(
      (acc, val, ind) => ind < 2 ? acc : [acc[1], acc[0] + acc[1]],
      [0, 1])

    return last2Fibs[0] + last2Fibs[1]
  }
}

module.exports = fib

File: fibonacci/python/forwardRecursiveWithReduce.py
from functools import reduce


def fib(n=None):
    if n is None or not str(n).isnumeric() or n < 0:
        return 'Invalid Input!'

    if n < 2:
        return n

    ret = reduce(lambda acc, val: acc if val < 2 else [acc[1], acc[0] + acc[1]], range(n), [0, 1])
    return ret[0] + ret[1]
